File: skript-parser-java/src/main/java/io/github/syst3ms/skriptparser/util/StringUtils.java
package io.github.syst3ms.skriptparser.util;

public class StringUtils {
	public static String getEnclosedText(String pattern, char opening, char closing, int start) {
	int n = 0;
	for (int i = start; i < pattern.length(); i++) {
		char c = pattern.charAt(i);
		if (c == '\\') {
			i++;
		} else if (c == closing) {
			n--;
			if (n == 0) {
				return pattern.substring(start + 1, i); // We don't want the beginning bracket in there
			}
		} else if (c == opening) {
			n++;
		}
	}
	return null;
}
}

File: skript-parser-kotlin/src/main/kotlin/io/github/syst3ms/skriptparser/util/StringUtils.kt
package io.github.syst3ms.skriptparser.util

fun String.getEnclosedText(opening: Char, closing: Char, start: Int = 0): String? {
    var n = 0
    var i = start
    while (i < length) {
        val c = this[i]
        if (c == '\\') {
            i++
        } else if (c == closing) {
            n--
            if (n == 0) {
                return substring(start + 1, i) // We don't want the beginning bracket in there
            }
        } else if (c == opening) {
            n++
        }
        i++
    }
    return null
}